sdks/java/harness/src/main/java/org/apache/beam/fn/harness/SplittableSplitAndSizeRestrictionsDoFnRunner.java

```diff
@@ -0,0 +1,481 @@
+
+  private final boolean observesWindow;
+  private final PipelineOptions pipelineOptions;
+
+  private final DoFnInvoker<InputT, OutputT> doFnInvoker;
+
+  private final FnDataReceiver<
+          WindowedValue<KV<KV<InputT, KV<RestrictionT, WatermarkEstimatorStateT>>, Double>>>
+      mainOutputConsumer;
+
+  private final FnApiStateAccessor<?> stateAccessor;
+  private final SplitRestrictionArgumentProvider mutableArgumentProvider;
+
+  private final DoFnSchemaInformation doFnSchemaInformation;
+  private final Map<String, PCollectionView<?>> sideInputMapping;
+
+  private RestrictionT currentRestriction;
+  private WatermarkEstimatorStateT currentWatermarkEstimatorState;
+
+  private RestrictionTracker<RestrictionT, PositionT> currentTracker;
+
+  /**
+   * If non-null, set to true after currentTracker has had a tryClaim issued on it. Used to ignore
+   * checkpoint split requests if no progress was made.
+   */
+  private @Nullable AtomicBoolean currentTrackerClaimed;
+
+  SplittableSplitAndSizeRestrictionsDoFnRunner(
+      PipelineOptions pipelineOptions,
+      PTransform pTransform,
+      RunnerApi.Components components,
+      Function<String, FnDataReceiver<WindowedValue<?>>> getPCollectionConsumer,
+      BundleFinalizer bundleFinalizer,
+      FnApiStateAccessor.Factory<Object> stateAccessorFactory)
+      throws IOException {
+    this.pipelineOptions = pipelineOptions;
+
+    ParDoPayload parDoPayload = ParDoPayload.parseFrom(pTransform.getSpec().getPayload());
+
+    // DoFn and metadata
+    DoFn<InputT, OutputT> doFn = (DoFn<InputT, OutputT>) ParDoTranslation.getDoFn(parDoPayload);
+    DoFnSignature doFnSignature = DoFnSignatures.signatureForDoFn(doFn);
+    this.doFnInvoker = DoFnInvokers.tryInvokeSetupFor(doFn, pipelineOptions);
+    this.doFnSchemaInformation = ParDoTranslation.getSchemaInformation(parDoPayload);
+
+    // Main output
+    checkArgument(
+        pTransform.getOutputsMap().size() == 1,
+        "SplitAndSizeRestrictions expects exact one output, but got: ",
+        pTransform.getOutputsMap().size());
+    TupleTag<KV<KV<InputT, KV<RestrictionT, WatermarkEstimatorStateT>>, Double>> mainOutputTag =
+        new TupleTag<>(Iterables.getOnlyElement(pTransform.getOutputsMap().keySet()));
+    @SuppressWarnings("rawtypes") // cannot do this multi-level cast without rawtypes
+    FnDataReceiver<
+            WindowedValue<KV<KV<InputT, KV<RestrictionT, WatermarkEstimatorStateT>>, Double>>>
+        mainOutputConsumer =
+            (FnDataReceiver)
+                getPCollectionConsumer.apply(pTransform.getOutputsMap().get(mainOutputTag.getId()));
+    this.mainOutputConsumer = mainOutputConsumer;
+
+    // Side inputs
+    this.sideInputMapping = ParDoTranslation.getSideInputMapping(parDoPayload);
+
+    // Register processing methods
+    this.observesWindow =
+        (doFnSignature.splitRestriction() != null
```

Comment:
could doFnSignature have a observesWindow() helper?

Reply:
Good thought. Looking into it, this would cause window-observing behavior in e.g. `@ProcessElement` to cause us to believe that `@SplitRestriction` also observes the window.